docs/proposals/cross-mesh-ux.md

```diff
@@ -0,0 +1,229 @@
+traffic.
+This option will allow us to configure the corresponding Envoy listener
+with all mesh CAs.
+
+```
+type: MeshGateway
+mesh: expose
+name: edge-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8090
+    protocol: HTTPS
+    hostname: expose.mesh # optional
+    tags:
+      name: mesh-listener
+    tls:
+      mode: TERMINATE
+      options:
+        crossMesh: true # new option
+```
+
+There are two ways we can enable configuring more specific access rules.
+
+#### `MeshGatewayTrafficPermission`
+
+The first is to add a _new resource_ that defines which
+cross-mesh traffic is allowed for a matching `MeshGateway` listener and
+`routeName`.
+
+```
+type: MeshGatewayTrafficPermission
+mesh: expose
+name: cross-mesh
+selector: # these match a listener
+- tags:
+    kuma.io/service: edge-gateway
+    name: mesh-listener
+default: deny
+rules:
+- routeName: cross-mesh
+  action: allow
+  sources:
+  - mesh:
+      name: consume # or '*' for any mesh
+      tags:
+        kuma.io/service: client
+```
+
+This concentrates `MeshGateway` permission management into its own resource.
+This may be a more generally useful abstraction given potential
+other kinds of `sources`. For example, one for filtering based on JWT token:
+
+```
+  - routeName: cross-mesh
+    action: allow
+    sources:
+    - user:
+        email: abc@xyz.com
+```
+
+#### Inline `MeshGatewayRoute`
+
+Another option is to add cross-mesh access control to the `MeshGatewayRoute`
+resource, as another type of `conf.http.rules.matches` entry or
+potentially at the top level under `conf.http`.
+
+```
+type: MeshGatewayRoute
+mesh: expose
+name: edge-gateway-route
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+      name: mesh-listener
+conf:
+  http:
+    mesh:                    # new entry type
+      - name: consume
+        tags:
+          kuma.io/service: client
+    rules:
+      - matches:
+          # ALL match types must succeed
+          - path:
+              match: PREFIX
+              value: /api
+          - mesh:                    # and/or a new entry type here
+            - name: consume
+              tags:
+                kuma.io/service: client
+        backends:
+          - destination:
+              kuma.io/service: server
+      - matches:    # ALL matches must succeed
+          - path:
+              match: PREFIX
+              value: /api/v2
+          - mesh:
+            - name: *
+        backends:
+          - destination:
+              kuma.io/service: server
+```
+
+The advantage of this option is probably reduced complexity for less complicated
+cases.
+
+### `consume` UX
+
+In order for `consume` services to communicate with `expose`, we need to give
+those services a way to contact the `MeshGateway`. A `MeshGateway` doesn't have
+an address or concrete identity beyond the identities of the `kuma-dp` proxies that
+serve the gateway.
+
+#### Virtual outbounds
```

Comment:
Can we use `ExternalService` to generate an outbound? Maybe it should be a special type of the external service like:
```yaml
type: ExternalService
mesh: consume
name: es-edge-gateway
networking:
  crossMesh:
    mesh: expose
    service: edge-gateway
```
It just feels like ExternalService is more natural way to express the desire of consuming the service from another mesh.

Reply:
We discussed in [a meeting about cross-mesh](https://docs.google.com/document/d/1oFF9N1hN_j5MI_JM0NHZOAjzVUV6ycYVkvk1bHlt0x4) the pros and cons of allowing a Mesh to define an API for how it can be consumed via other Meshes as opposed to allowing service to service traffic.

Reply:
Ah ok. Well, I don't think it's a service we're consuming really, it's the `MeshGateway`.

I wonder if we can just avoid an explicit `VirtualOutbound` or `ExternalService` entirely and just use the `.hostname` on the listener to generate a name.

Reply:
Added this to the proposal.